server/src/main/java/org/elasticsearch/indices/IndicesWriteLoadStatsCollector.java

```diff
@@ -0,0 +1,242 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.IndexEventListener;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.LongSupplier;
+import java.util.stream.StreamSupport;
+
+public class IndicesWriteLoadStatsCollector implements IndexEventListener, ClusterStateListener {
+    private final ClusterService clusterService;
+    private final IndicesService indicesService;
+    private final LongSupplier relativeTimeInNanosSupplier;
+    private final Map<ShardId, ShardWriteLoadHistogram> histograms = ConcurrentCollections.newConcurrentMap();
+    private final AtomicBoolean clusterStateReady = new AtomicBoolean();
+
+    public IndicesWriteLoadStatsCollector(
+        ClusterService clusterService,
+        IndicesService indicesService,
+        LongSupplier relativeTimeInNanosSupplier
+    ) {
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
+        this.relativeTimeInNanosSupplier = relativeTimeInNanosSupplier;
+        clusterService.addListener(this);
+    }
+
+    void collectWriteLoadStats() {
+        for (IndexShard indexShard : getDataStreamsWriteIndicesShards()) {
+            final var parentDataStreamName = getParentDataStreamName(indexShard.shardId().getIndexName());
+            final var shardWriteLoadHistogram = histograms.computeIfAbsent(
+                indexShard.shardId(),
+                shardId -> new ShardWriteLoadHistogram(
+                    parentDataStreamName,
+                    shardId,
+                    indexShard.routingEntry().primary(),
+                    relativeTimeInNanosSupplier
+                )
+            );
+            long totalIndexingTime = Math.addExact(indexShard.getTotalIndexingTimeInNanos(), indexShard.getTotalDeleteTimeInNanos());
+            long totalMergeTimeInMillis = indexShard.getTotalMergeTimeInMillis();
+            long totalRefreshTimeInNanos = indexShard.getTotalRefreshTimeInNanos();
+            shardWriteLoadHistogram.recordSample(totalIndexingTime, totalMergeTimeInMillis, totalRefreshTimeInNanos);
+        }
+    }
+
+    Iterable<IndexShard> getDataStreamsWriteIndicesShards() {
+        // Wait until the node has access to the cluster state
+        if (clusterStateReady.get() == false) {
+            return Collections.emptyList();
+        }
+
+        return () -> clusterService.state()
+            .metadata()
+            .getIndicesLookup()
+            .values()
```

Comment:
This looks sort of expensive in a many-indices world? I wonder if iterating `IndicesService` instead is cheaper? Or perhaps maintaining a list of the relevant shards? We almost have that in `histograms` anyway.